File: sort/sort.cpp
//
// Created by spendar on 2022/3/14.
//
#include "sort.h"


void Sort::bubbleSort(std::vector<int>& arr) {
	int size = arr.size();
	for(int i = 0; i < size; i++) {
		for(int j = 0; j < size-i-1; j++) {
			if(arr[j] > arr[j+1]) {
				int temp = arr[j];
				arr[j] = arr[j+1];
				arr[j+1] = temp;
			}
		}
	}
}

void Sort::selectSort(std::vector<int>& arr) {
	int size = arr.size();
	for(int i = 0; i < size; i++) {
		int index = 0;
		int temp = arr[0];
		for(int i = 0; i < size-i; i++) {
			if(arr[i] > temp) {
				index = i;
				temp = arr[i];
			}
		}
		arr[size-1-i] = arr[index]; 
		arr[index] = temp;
	}
}

void Sort::insertSort(std::vector<int>& arr) {
	int size = arr.size();
	for(int i = 1; i < size; i++) {
		int temp = arr[i];
		int index = i;
		while(index > 0 && arr[index-1] < temp) {
			arr[index] = arr[index-1];
		}
		arr[index] = temp;
	}
}

void Sort::shellSort(std::vector<int>& arr) {
	int size = arr.size();
	for(int gap = size/2; gap > 0; gap /= 2){
		for(int i = gap; i < size; i++){
			int index = i-gap;
			int temp = arr[i];
			while(index > 0 && temp < arr[index]) {
				arr[index+gap] = arr[index];
				index -= gap;
			}
			arr[index+gap] = temp;
		}
	}
}

void merge(std::vector<int>& arr, int left, int right) {
	if(left >= right) {
		return;
	}
	int mid = (right - left) / 2 + left;
	merge(arr, left, mid);
	merge(arr, mid+1, right);
	int temp[right-left+1];
	int index = 0;
	int l = left;
	int r = mid + 1;
	while(l <= mid && r <= right) {
		if(arr[l] <= arr[r]) {
			temp[index++] = arr[l++];
		} else {
			temp[index++] = arr[r++];
		}
	}
	while(l <= mid) {
		temp[index++] = arr[l++];
	}
	while(r <= right) {
		temp[index++] = arr[r++];
	}
	for(int i = left; i <= right; i++) {
		arr[i] = temp[i-left];
	}
}

void Sort::mergeSort(std::vector<int>& arr) {
	merge(arr, 0, arr.size()-1);
}

void Sort::heapSort(std::vector<int>& arr) {
	std::priority_queue<int> pq;
	for(int a : arr) {
		pq.push(a);
	}
	int size = arr.size();
	for(int i = size-1; i >= 0; i--) {
		arr[i] = pq.top();
		pq.pop();
	}
}

void quick(std::vector<int>& arr, int left, int right) {
	if(left >= right) {
		return;
	}
	int key = arr[left];
	int l = left;
	int r = right;
	while(l < r) {
		while(l < r && arr[r] >= key) {
			r--;
		}
		if(l < r) {
			arr[l] = arr[r];
		}
		while(l < r && arr[l] <= key) {
			l++;
		}
		if(l < r) {
			 arr[r] = arr[l];
		}
	}
	arr[l] = key;
	quick(arr, left, l-1);
	quick(arr, l+1, right);
}

void Sort::quickSort(std::vector<int>& arr) {
	quick(arr, 0, arr.size()-1);
}

void Sort::countSort(std::vector<int>& arr) {
	int maxn = 0x80000000;
	int minn = 0x7fffffff;
	for(int a : arr) {
		maxn = std::max(maxn, a);
		minn = std::min(minn, a);
	}
	int data[maxn - minn+1];
	int index = 0;
	for(int i = 0; i < maxn - minn+1; i++) {
		data[i] = 0;
	}
	for(int a : arr) {
		data[a-minn]++;
	}
	for(int i = 0; i < maxn - minn+1; i++) {
		for(int j = 0; j < data[i]; j++) {
			arr[index++] = i + minn;
		}
	}
}

void Sort::bucketSort(std::vector<int>& arr) {
	int nums_max = arr[0];
	int nums_min = arr[0];
	for(int i = 0; i < arr.size(); i++){
		if(arr[i] < nums_min)
			nums_min = arr[i];
		if(arr[i] > nums_max)
			nums_max = arr[i];
	}
	int gap = (nums_max - nums_min) / 5 + 1;
	std::vector<int> buckets[5];
	for(int i = 0; i < arr.size(); i++)
		buckets[(arr[i] - nums_min) / gap].push_back(arr[i]);
	for(int i = 0; i < 5; i++){
		quick(buckets[i], 0, buckets[i].size()-1);
	}
	int index = 0;
	for(int i = 0; i < 5; i++) {
		for(int j = 0; j < buckets[i].size(); j++) {
			arr[index++] = buckets[i][j];
		}
	}
}

void Sort::radixSort(std::vector<int>& arr) {
	int size = arr.size();
	int nums_max = arr[0];
	for(int i = 0; i < size; i++)
		if(nums_max < arr[i])
			nums_max = arr[i];
	int radix = 1;
	while(nums_max /= 10)
		radix++;
	for(int k = 0; k < radix; k++){
		std::vector<int> radixes[10];
		for(int i = 0; i < size; i++)
			radixes[arr[i]/((int)(pow(10, k))%10)].push_back(arr[i]);
		int index = 0;
		for(int i = 0; i < 10; i++)
			for(int j = 0; j < radixes[i].size(); j++)
				arr[index++] = radixes[i][j];
		}
}

File: sort/Sort.java
package com.company;

public class Sort {

    /**
     * 冒泡排序
     * @param a 待排序数组
     */
    public void bubbleSort(int[] a){
        int len = a.length;
        for(int i = 0; i < len; i++){
            for(int j = 1; j < len-i; j++){
                if(a[j] < a[j-1]){
                    int temp = a[j];
                    a[j] = a[j-1];
                    a[j-1] = temp;
                }
            }
        }
    }

    /**
     * 选择排序
     * @param a 待排序数组
     */
    public void selectSort(int[] a){
        int len = a.length;
        for(int i = 0; i < len; i++){
            int index = 0;
            int temp = a[0];
            for(int j = 0; j < len-i; j++){
                if(a[j] > temp){
                    index = j;
                    temp = a[j];
                }
            }
            int t = a[len-1-i];
            a[len-1-i] = temp;
            a[index] = t;
        }
    }

    /**
     * 插入排序
     * @param a 待排序数组
     */
    public void insertSort(int[] a){
        int len = a.length;
        for(int i = 1; i < len; i++){
            int temp = a[i];
            int index = i;
            while(index >= 1 && temp < a[index-1]) {
                a[index] = a[index - 1];
                index--;
            }
            a[index] = temp;
        }
    }

    /**
     * 快速排序
     * @param a 待排序数组
     * @param start 数组起始位置
     * @param end 数组结束位置，左闭右开
     */
    public void quickSort(int[] a, int start, int end){
        if(start >= end)    return;
        int index = partion(a, start, end);
        quickSort(a, start, index);
        quickSort(a, index+1, end);
    }

    private int partion(int[] a, int start, int end) {
        int key = a[start];
        int left = start;
        int right = end - 1;
        while(left < right){
            while(left < right && a[right] > key)
                right--;
            if(left < right){
                int temp = a[right];
                a[right] = a[left];
                a[left] = temp;
            }
            while(left < right && a[left] < key)
                left++;
            if(left < right){
                int temp = a[right];
                a[right] = a[left];
                a[left] = temp;
            }
        }
        return left;
    }

    /**
     * 归并排序
     * @param a 待排序数组
     * @param start 数组起始位置
     * @param end 数组结束位置 左闭右开
     */
    public void mergeSort(int a[], int start, int end){
        if(start>= end-1) return;
        int mid = (start + end) / 2;
        mergeSort(a, start, mid);
        mergeSort(a, mid, end);
        merge(a, start, mid, end);
    }

    private void merge(int a[], int start, int mid, int end){
        int t[] = new int[end - start];
        int index = 0;
        int index1 = start;
        int index2 = mid;
        while(index1 < mid && index2 < end){
            if(a[index1] < a[index2])
                t[index++] = a[index1++];
            else
                t[index++] = a[index2++];
        }
        while(index1 < mid)
            t[index++] = a[index1++];
        while(index2 < end)
            t[index++] = a[index2++];
        for(int i = 0; i < end-start; i++)
            a[start+i] = t[i];
    }

    /**
     * 堆排序
     * @param a 待排序数组
     */
    public void headSort(int a[]){
        Heap heap = new Heap(a.length, false);
        for(int n : a)
            heap.push(n);
        for(int i = 0; i < a.length; i++)
            a[i] = heap.pop();
    }

    public static void main(String[] args){
        int a[] = {6,8,3,7,2,0,9,5,4,1};
        Sort s = new Sort();
        s.headSort(a);
        for(int num: a)
            System.out.print(num + " ");
    }
}

